File: Hashing/python/HashTable.py
class HashTable:
    # Initialize the hash table with a default capacity
    def __init__(self, capacity=10):
        self.numBuckets = capacity
        self.bucketArray = [[] for _ in range(capacity)]  # List of lists to store chains
        self.size = 0  # Number of key-value pairs in the hash table

    # Hash function to calculate the index for a given key
    def _hash(self, key):
        return key % self.numBuckets

    # Method to add a key-value pair to the hash table
    def put(self, key, value):
        bucketIndex = self._hash(key)  # Calculate the index in the bucket array
        bucket = self.bucketArray[bucketIndex]  # Get the list at the index (chain)
        
        for pair in bucket:
            if pair[0] == key:  # If the key already exists, update its value
                pair[1] = value
                return
        
        bucket.append([key, value])  # If the key doesn't exist, add it
        self.size += 1  # Increment the size of the hash table

    # Method to retrieve the value associated with a key
    def get(self, key):
        bucketIndex = self._hash(key)  # Calculate the index in the bucket array
        bucket = self.bucketArray[bucketIndex]  # Get the list at the index (chain)
        
        for pair in bucket:
            if pair[0] == key:  # If the key is found, return its value
                return pair[1]
        
        return None  # Return None if the key is not found

    # Method to remove a key-value pair from the hash table
    def remove(self, key):
        bucketIndex = self._hash(key)  # Calculate the index in the bucket array
        bucket = self.bucketArray[bucketIndex]  # Get the list at the index (chain)
        
        for i, pair in enumerate(bucket):
            if pair[0] == key:  # If the key is found, remove it from the list
                del bucket[i]
                self.size -= 1  # Decrement the size of the hash table
                return pair[1]  # Return the removed value
        
        return None  # Return None if the key is not found

    # Method to return the number of key-value pairs in the hash table
    def size(self):
        return self.size

    # Method to check if the hash table is empty
    def is_empty(self):
        return self.size == 0

# Example usage
if __name__ == "__main__":
    table = HashTable(10)
    table.put(105, "Tom")
    table.put(21, "Sana")
    table.put(31, "John")
    
    print(table.size)      # Outputs: 3
    print(table.get(31))   # Outputs: "John"
    print(table.get(88))   # Outputs: None (since key 88 does not exist)
    print(table.remove(21))# Outputs: "Sana"
    print(table.size)      # Outputs: 2
    print(table.get(21))   # Outputs: None (since key 21 was removed)

File: Hashing/java/HashTable.java
package Hashing.java;

// Implementing HashTable using Separate Chaining (Linked Lists)
public class HashTable{
  private HashNode[] bucketArray; // Array to hold linked lists (chains)
  private int numBuckets;         // Capacity of the HashTable
  private int size;               // Number of key-value pairs in the HashTable

  public HashTable() {
    this(10); // Default capacity is 10 if no capacity is provided
  }

  public HashTable(int capacity) {
    this.numBuckets = capacity;
    bucketArray = new HashNode[numBuckets];
    this.size = 0;
  }

  // HashNode represents each entry in the HashTable
  private class HashNode {
    private Integer key; // Key of the HashNode
    private String value; // Value of the HashNode
    private HashNode next; // Pointer to the next node in the chain

    public HashNode(Integer key, String value) {
      this.key = key;
      this.value = value;
    }
  }

  // Methods

  // Returns the number of key-value pairs in the HashTable
  public int size() {
    return size;
  }

  // Checks if the HashTable is empty
  public boolean isEmpty() {
    return size == 0;
  }

  // Computes the index for a given key (using a hash function)
  private int getBucketIndex(Integer key) {
    return key % numBuckets; // Hash function using modulus
  }

  // Adds a key-value pair to the HashTable
  public void put(Integer key, String value) {
    if(key == null || value == null) {
      throw new IllegalArgumentException("Key or Value cannot be null");
    }

    int bucketIndex = getBucketIndex(key); // Get index in the bucket array
    HashNode head = bucketArray[bucketIndex]; // Get the head of the chain at the bucket index

    // Check if the key already exists; if so, update its value
    while(head != null) {
      if(head.key.equals(key)) {
        head.value = value;
        return;
      }
      head = head.next;
    }

    // Insert the new key-value pair at the beginning of the chain
    size++;
    head = bucketArray[bucketIndex];
    HashNode newNode = new HashNode(key, value);
    newNode.next = head;
    bucketArray[bucketIndex] = newNode;
  }

  // Retrieves the value associated with a key
  public String get(Integer key) {
    int bucketIndex = getBucketIndex(key); // Get index in the bucket array
    HashNode head = bucketArray[bucketIndex]; // Get the head of the chain at the bucket index

    // Search the chain for the key
    while (head != null) {
      if(head.key.equals(key)) {
        return head.value; // Return the value if the key is found
      }
      head = head.next;
    }
    return null; // Return null if the key is not found
  }

  // Removes a key-value pair from the HashTable
  public String remove(Integer key) {
    int bucketIndex = getBucketIndex(key); // Get index in the bucket array
    HashNode head = bucketArray[bucketIndex]; // Get the head of the chain at the bucket index
    HashNode prev = null; // To keep track of the previous node

    // Search the chain for the key
    while(head != null) {
      if(head.key.equals(key)) {
        break; // Stop if the key is found
      }
      prev = head;
      head = head.next;
    }

    if(head == null) return null; // If key is not found, return null

    size--; // Decrement size as we are removing an entry

    if(prev != null) {
      prev.next = head.next; // Bypass the node to remove it from the chain
    } else {
      bucketArray[bucketIndex] = head.next; // Remove the head of the chain
    }

    return head.value; // Return the removed value
  }

  public static void main(String[] args) {
    HashTable table = new HashTable(10);
    table.put(105, "Tom");
    table.put(21, "Sana");
    table.put(31, "John");

    System.out.println(table.size());    // Outputs: 3
    System.out.println(table.get(31));   // Outputs: "John"
    System.out.println(table.get(88));   // Outputs: null (since key 88 does not exist)
  }
}
